### ISSUE TITLE
Command 'None' in image 'my-docker-image' returned non-zero exit status when running on Windows

### ISSUE THREAD

**Issue by vmenger** [MrMino/dockernel#7]
Hi, 

Very nice project. I'm trying to run on Windows using the example I found here: 

https://stackoverflow.com/questions/63702536/can-you-make-jupyter-start-up-a-kernel-in-a-docker-container/63715102#63715102

Dockerfile contents: 

```
FROM python:3.7-slim-buster

RUN pip install --upgrade pip ipython ipykernel
CMD python -m ipykernel_launcher -f $DOCKERNEL_CONNECTION_FILE
```

Install command: 

```
docker build --tag my-docker-image /path/to/the/dockerfile/dir
pip install dockernel
dockernel install my-docker-image
```

This part works fine, but when I try to create a jupyter notebook, it cannot connect to kernel. From the jupyter notebook terminal:

```
[I 11:51:23.235 NotebookApp] Kernel started: dc7e2d74-9605-4ebf-a093-cacbce3035f2, name: c3a0179b8100
Traceback (most recent call last):
  File "C:\Users\VincentMenger\miniconda3\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "C:\Users\VincentMenger\miniconda3\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\Users\VincentMenger\miniconda3\lib\site-packages\dockernel\__main__.py", line 10, in <module>
    sys.exit(main())
  File "C:\Users\VincentMenger\miniconda3\lib\site-packages\dockernel\__main__.py", line 6, in main
    return run(sys.argv)
  File "C:\Users\VincentMenger\miniconda3\lib\site-packages\dockernel\app.py", line 7, in run
    return run_subcommand(parsed_args)
  File "C:\Users\VincentMenger\miniconda3\lib\site-packages\dockernel\cli\main.py", line 18, in run_subcommand
    return parsed_args.func(parsed_args)
  File "C:\Users\VincentMenger\miniconda3\lib\site-packages\dockernel\cli\start.py", line 63, in start
    stderr=True
  File "C:\Users\VincentMenger\miniconda3\lib\site-packages\docker\models\containers.py", line 839, in run
    container, exit_status, command, image, out
docker.errors.ContainerError: Command 'None' in image 'my-docker-image' returned non-zero exit status 1
[I 11:51:26.230 NotebookApp] KernelRestarter: restarting kernel (1/5), new random ports
```

After which the same message is repeated 4 more times.

I also noticed a folder that is literally called `{connection_file}` created in my windows home folder. 

Is this a Windows compatibility issue? Or does the dockerfile not contain the appropriate commands? Happy to help solve the issue / debug some things on Windows but not sure where this issue arises. Any help is much appreciated.

**Comment by MrMino** [comment]
Hi

I don't have access to my Windows machine currently, but I'll check it as soon as I do. `{connection_file}` looks like I forgot to add a f-modifier to a string somewhere, definitely a bug.

**Comment by MrMino** [comment]
I've checked it on Windows - indeed, there were some bugs that prevented `dockernel start` from working properly.

All of the fixes are going onto `next-patch` branch.

There are two large issues to be tackled right now:

1. FUSE sometimes refuses to operate properly on non-edge versions of docker for-win: https://github.com/docker/for-win/issues/5955. By comment from @stephen-turner, this is fixed in 2.3.2.0.
2. IPython kernels require connection spec files to have `0600` permissions set, whereas docker for-win mounts files in bind mode with `0755`, which is currently unchangeable.

The first one can be fixed by using the edge version of Docker Desktop. The second one I'm working on currently. There are some weird hacks involved in this, as getting around this limitation requires creating something akin to a cross-device hardlink in the container - which is impossible with current kernels.

I'll update you as soon as I get it done.

**Comment by vmenger** [comment]
Thanks, I really appreciate the efforts. Will look into your reply in more detail when I find the time.

**Comment by RoelantStegmann** [comment]
Also curious how this issue is going :) Dockernel looks cool, would be happy to use it :)

**Comment by vmenger** [comment]
@MrMino Finally had some time to check up on this. I tried your suggestion (adding `cp` to the image command). iPython does modify the file, but only the `kernel_name` field which as far as I can tell is not doing any harm. It solves the permission error, and docker seems to digest everything correctly, however, the notebook cannot connect to the kernel. 

Docker output: 

```
NOTE: When using the ipython kernel entry point, Ctrl-C will not work.


To exit, you will have to explicitly quit this process, by either sending

"quit" from a client, or using Ctrl-\ in UNIX-like environments.


To read more about this, see https://github.com/ipython/ipython/issues/2049



To connect another client to this kernel, use:

--existing /kernel-connection-spec.json
```


Jupyter notebook output: 

```
[W 19:12:20.854 NotebookApp] Replacing stale connection: 110b40b9-bae7-4f58-819f-305b2a4fa2f5:7c3c1c10a8de48d083b7b684482786f4
[W 19:12:42.951 NotebookApp] Replacing stale connection: 110b40b9-bae7-4f58-819f-305b2a4fa2f5:7c3c1c10a8de48d083b7b684482786f4
[W 19:12:59.878 NotebookApp] Timeout waiting for kernel_info reply from 110b40b9-bae7-4f58-819f-305b2a4fa2f5
[I 19:12:59.894 NotebookApp] Starting buffering for 110b40b9-bae7-4f58-819f-305b2a4fa2f5:7c3c1c10a8de48d083b7b684482786f4
[I 19:12:59.894 NotebookApp] Restoring connection for 110b40b9-bae7-4f58-819f-305b2a4fa2f5:7c3c1c10a8de48d083b7b684482786f4
```

I suspect maybe some networking-related issue, but fiddling around with the network mode and publishing ports did not resolve the anything. Do you have any other pointers or thoughts upon reading this?

**Comment by vmenger** [comment]
BTW, this command resolves the permission error as well on my system and keeps the binding intact: 

`CMD chmod 600 $DOCKERNEL_CONNECTION_FILE; python -m ipykernel_launcher -f $DOCKERNEL_CONNECTION_FILE`

**Comment by vmenger** [comment]
Ah, did some more debugging. It works if you set `"ip": "0.0.0.0"` in the connection file, set `network_mode=bridge` and explicitly add the ports with `ports={shell_port:shell_port, ...}`. 

I can do a PR later if that is helpful?

**Comment by lwhite-decentrix** [comment]
@vmenger Trying to work through this solution for Windows. How do you set ip in the connection file?

**Comment by vmenger** [comment]
I fixed it like this (ignore the mounts and the auto_remove=False):

https://github.com/vmenger/dockernel/commit/395ae0859209b0d24473f6d3a340a8bb5d12b8a7

**Comment by MrMino** [comment]
Should be fixed with the release of `dockernel==1.0.2`, but if the issue persist - please don't hesitate to ping me.
